### ISSUE TITLE
tf2.0 takes more than twice memory than 1.14

### ISSUE THREAD

**Issue by saintthor** [tensorflow/tensorflow#33412]
I have 2 environments. tf 1.14 installed for python 3.5 and tf 2 for python 3.7.

I run the same code and load the same keras model in both environments. in py3.5, after model loaded, the process takes 560MB memory. while in py3.7, it takes 1.2GB.

is there any way to reduce the memory for tf 2 ?

**Comment by mihaimaruseac** [comment]
Could you try providing a reproducer?

**Comment by ravikyram** [comment]
@saintthor 
In order to expedite the trouble-shooting process, please provide a minimal standalone code to reproduce the issue reported here. Thanks!

**Comment by saintthor** [comment]
> @saintthor
> In order to expedite the trouble-shooting process, please provide a minimal standalone code to reproduce the issue reported here. Thanks!

I tested with a simple code as:

`
import tensorflow as tf
from time import sleep
Model = tf.keras.models.load_model( 'models/mo_fghijk' )
print( 'loaded' )
sleep( 30 )
`

It takes 13.9% memory run by py3.7 and 10.8% by py3.5. the gap is not so much as running my real code. but it is obvious.

**Comment by saintthor** [comment]
with psutil:
`import psutil
import os
print( 'start', psutil.Process(os.getpid()).memory_info())
import tensorflow as tf
print( 'imported', psutil.Process(os.getpid()).memory_info())
Model = tf.keras.models.load_model( 'models_1/mo_fghijk' )
print( 'loaded', psutil.Process(os.getpid()).memory_info())
`
py3.5
start pmem(rss=12574720, vms=46747648)
imported pmem(rss=234987520, vms=1003446272)
loaded pmem(rss=442376192, vms=2402095104)

py3.7
start pmem(rss=11087872, vms=59379712, shared=5435392, text=2658304, lib=0, data=5623808, dirty=0)
imported pmem(rss=195751936, vms=701423616, shared=90480640, text=2658304, lib=0, data=224903168, dirty=0)
loaded pmem(rss=573812736, vms=2255093760, shared=98611200, text=2658304, lib=0, data=1778573312, dirty=0)

**Comment by saintthor** [comment]
In my real code, the model is not used frequently. I get the data once every 5 minutes, predict, then sleep to next time.
The model file stored on the disk is only 4MB. I hope there is a minimum loading mode to save the memory even run slower.

**Comment by ymodak** [comment]
@saintthor Thanks sharing the memory usage stats. However at this point we cannot reproduce the reported behavior since we don't have access to your local model.  
Perhaps you can build a similar model using predefined [tf.keras.datasets](https://www.tensorflow.org/api_docs/python/tf/keras/datasets) or attach your model if possible. Thanks!

**Comment by saintthor** [comment]
> @saintthor Thanks sharing the memory usage stats. However at this point we cannot reproduce the reported behavior since we don't have access to your local model.
> Perhaps you can build a similar model using predefined [tf.keras.datasets](https://www.tensorflow.org/api_docs/python/tf/keras/datasets) and see if you can repro the bug or attach your model if possible. Thanks!

thanks. please test with this file.

[model.zip](https://github.com/tensorflow/tensorflow/files/3756628/model.zip)

**Comment by ymodak** [comment]
I was able to reproduce reported behavior TF 2.X consumes more than twice the memory taken by TF 1.X

For TF '1.15.0'
```python
start pmem(rss=407224320, vms=1446768640, shared=173322240, text=3883008, lib=0, data=493367296, dirty=0)
imported pmem(rss=407224320, vms=1446768640, shared=173322240, text=3883008, lib=0, data=493367296, dirty=0)
```

For TF '2.1.0-dev20191029'
```python
start pmem(rss=157614080, vms=665616384, shared=63725568, text=3883008, lib=0, data=213745664, dirty=0)
imported pmem(rss=433348608, vms=2544324608, shared=222384128, text=3883008, lib=0, data=393576448, dirty=0)
loaded pmem(rss=1765462016, vms=8929935360, shared=396664832, text=3883008, lib=0, data=1685512192, dirty=0)
```

**Comment by saintthor** [comment]
> I was able to reproduce reported behavior TF 2.X consumes more than twice the memory taken by TF 1.X
> 
> For TF '1.15.0' [GitHub Gist](https://colab.sandbox.google.com/gist/ymodak/4d66c77e8c807b8da3614d6919da6cd2/untitled18.ipynb)
> 
> ```python
> start pmem(rss=407224320, vms=1446768640, shared=173322240, text=3883008, lib=0, data=493367296, dirty=0)
> imported pmem(rss=407224320, vms=1446768640, shared=173322240, text=3883008, lib=0, data=493367296, dirty=0)
> ```
> 
> For TF '2.1.0-dev20191029' [GitHub Gist](https://colab.sandbox.google.com/gist/ymodak/a91ea97d92eca586ac8c577db3c96fe5/untitled19.ipynb#scrollTo=x_pzvzWWTI3S)
> 
> ```python
> start pmem(rss=154517504, vms=658960384, shared=62566400, text=3883008, lib=0, data=211365888, dirty=0)
> imported pmem(rss=431603712, vms=2538975232, shared=221831168, text=3883008, lib=0, data=392503296, dirty=0)
> loaded pmem(rss=604594176, vms=2801840128, shared=235962368, text=3883008, lib=0, data=636260352, dirty=0)
> ```
thanks.

For TF '1.15.0'，the start and imported logs seems same. because you imported tensorflow at the beginning. We may mainly compare the imported and loaded logs.

**Comment by KichangKim** [comment]
Related issue with reproducible code : #31871

**Comment by karmel** [comment]
In 2.0, we have extended the loading and saving functionality to cover a number of use cases and to use a different format, so this is not entirely surprising. Can you clarify how the model is being saved in each case? Is the saved file the same size in each case?

**Comment by saintthor** [comment]
> In 2.0, we have extended the loading and saving functionality to cover a number of use cases and to use a different format, so this is not entirely surprising. Can you clarify how the model is being saved in each case? Is the saved file the same size in each case?

I saved the model by tf.keras.Model.save method.

I loaded the same model file in the two envs for the above testing.  thanks.

**Comment by MLnick** [comment]
We are facing a similar issue with some LSTM-based models we are working with. For our full model architecture, we are seeing memory usage of `2.8GB` for TF 2.x, vs `1.4GB` for TF 1.15. Combined with the memory leak issue [here](https://github.com/tensorflow/tensorflow/issues/40171), this makes TF 2.x unusable for our production models - both due to significantly higher memory usage (and cost) and the inability to delete/reload models due to the memory leaks.

I've created a simple reproduction to illustrate the memory consumption issue:

### Model load 

* simple model architecture - single `Bidirectional LSTM` layer -> `Dense` layer -> `Softmax Activation`. Save model as SavedModel (no training, just initialized weights).
* loading this model into memory from SavedModel takes up to **17x** more memory in TF 2.x vs TF 1.15.
  * Colab notebook: [load SavedModel TF 1.15](https://colab.research.google.com/drive/1ZLFvDZr7cVA8m8DTQxZWR-vnPB5ah7eL?usp=sharing) - memory used **14MB**
  * Colab notebook: [load SavedModel TF 2.2](https://colab.research.google.com/drive/1Q7Dx-NQo9F2Ghbrmo77-l9jL_eF8VaO1?usp=sharing) - memory used **249MB**
* size of SavedModel on disk is also around **17x** more for TF 2.x vs TF 1.15 (also shown in the notebooks above).
* I confirmed the same order of magnitude differences for loading the SavedModels in TF-Serving.

### Model save

It's worth noting that saving this model as SavedModel also sees memory usage of similar magnitude (and differential between versions) to the model load scenario.

The following notebooks illustrate this (and can be used to re-create the SavedModel files used in the model load notebooks above):

* Colab notebook: [save model to SavedModel TF 1.15](https://colab.research.google.com/drive/1vV3hH2gUCdFR44sioNmpmDOh5xwxr95K?usp=sharing) - memory used **30MB**
* Colab notebook: [save model to SavedModel TF 2.2](https://colab.research.google.com/drive/1QJvxORQSuG4DnRBd6pMRfWsQRKKP08n8?usp=sharing) - memory used **339MB**

### General issue

While I've shown the LSTM case here (as the memory usage differential is particularly large), this also applies to even the simplest models (such as 2 or 3 Dense layers), where I saw memory usage (and size on disk) around 2-5x more in TF 2.x vs TF 1.15.

cc @frreiss @kmh4321

**Comment by MLnick** [comment]
@karmel @ymodak @mihaimaruseac any chance to take another look at this issue? Above ^^ I've provided more info and repro Colab notebooks. The memory usage difference seems particularly large for LSTM layers and this is impacting our production use cases for LSTM models.

**Comment by MLnick** [comment]
@karmel @ymodak @mihaimaruseac ping on the above - any idea to move this forward?

**Comment by hatooku** [comment]
+1, we are also facing @MLnick's issue. This makes TF 2.x unusable for us as well in TF Serving due to the high memory usage.

**Comment by sachinprasadhs** [comment]
There seems to be a significant improvement in thein imported prem and loaded prem TF 2.x code after testing it in  Tensorflow 2.7, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/8ceadb513b0f722ad1cc31da44fc0e73/33412.ipynb) and confirm. Thanks!

**Comment by saintthor** [comment]
> 

Yes, In tf2.7
